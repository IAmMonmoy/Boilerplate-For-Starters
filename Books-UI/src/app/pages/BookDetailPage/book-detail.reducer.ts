import {
  REQUEST,
  SUCCESS,
  FAILURE,
} from '../../shared/reducers/action-type.util';
import axios from 'axios';

export const ACTIONS_TYPES = {
  FETCH_BOOK: 'bookDetailPage/FETCH_BOOK',
};

export interface IBookDetailPageBook {
  bookId: string;
  title: string;
  year: string;
  description: string;
}

const initialState = {
  loading: false,
  errorMessage: null,
  book: [] as ReadonlyArray<IBookDetailPageBook>,
};

const apiUrl = 'book/';

export type BookDeatilPageState = Readonly<typeof initialState>;

const reducer = (
  state: BookDeatilPageState = initialState,
  action,
): BookDeatilPageState => {
  switch (action.type) {
    case REQUEST(ACTIONS_TYPES.FETCH_BOOK):
      return {
        ...state,
        loading: true,
      };
    case SUCCESS(ACTIONS_TYPES.FETCH_BOOK):
      return {
        ...state,
        loading: false,
        book: action.payload,
      };
    case FAILURE(ACTIONS_TYPES.FETCH_BOOK):
      return {
        ...state,
        loading: false,
        errorMessage: action.payload,
      };
    default:
      return state;
  }
};

export const getBook = id => {
  const requestUrl = apiUrl + `?id=${id}`;
  return {
    type: ACTIONS_TYPES.FETCH_BOOK,
    payload: axios.get(requestUrl),
  };
};

export default reducer;
