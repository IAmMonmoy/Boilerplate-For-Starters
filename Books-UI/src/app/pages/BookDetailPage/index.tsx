import React, { useEffect } from 'react';
import { Helmet } from 'react-helmet-async';
import Book from '../../components/Book';
import { Container, Row, Col } from 'react-bootstrap';
import { RouteComponentProps } from 'react-router';
import { getBook } from './book-detail.reducer';
import { IRootState } from 'app/shared/reducers';
import { connect } from 'react-redux';

export interface IBookDetailPageProps
  extends StateProps,
    DispatchProps,
    RouteComponentProps<any> {}

export const BookDetailPage = ({
  getBook,
  book,
  match,
}: IBookDetailPageProps) => {
  useEffect(() => {
    getBook(match.params.id);
  }, [getBook, match.params.id]);
  return (
    <Container>
      <Helmet>
        <title>Book Detail Page</title>
        <meta
          name="description"
          content="A React Boilerplate application homepage"
        />
      </Helmet>
      <Row>
        <Col>
          <Book
            id={book.bookId}
            title={book.title}
            year={book.year}
            details={book.description}
            size="lg"
          />
        </Col>
      </Row>
    </Container>
  );
};

const mapDispatchToProps = { getBook };

const mapStateToProps = (storeState: IRootState) => ({
  loading: storeState.bookDeatilPage.loading,
  book: storeState.bookDeatilPage.book.length
    ? storeState.bookDeatilPage.book[0]
    : { bookId: '', title: '', year: '', description: '' },
});

type StateProps = ReturnType<typeof mapStateToProps>;
type DispatchProps = typeof mapDispatchToProps;

export default connect(mapStateToProps, mapDispatchToProps)(BookDetailPage);
