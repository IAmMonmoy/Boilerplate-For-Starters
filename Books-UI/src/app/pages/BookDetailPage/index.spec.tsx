import React from 'react';
import { createRenderer } from 'react-test-renderer/shallow';
import configureStore from 'redux-mock-store';
import promiseMiddleware from 'redux-promise-middleware';
import thunk from 'redux-thunk';
import BookDetailPage from './index';
import { Provider } from 'react-redux';
const shallowRenderer = createRenderer();

let [getBook] = new Array(1).fill(jest.fn());
export const testableComponent = component => {
  const mockStore = configureStore([thunk, promiseMiddleware]);
  const store = mockStore({});
  return <Provider store={store}>{component}</Provider>;
};

describe('<BookDeatilPage/>', () => {
  it('should match the snapshot', () => {
    const component = testableComponent(<BookDetailPage {...shallowSetup()} />);
    const renderedOutput = shallowRenderer.render(component);
    expect(renderedOutput).toMatchSnapshot();
  });
});

function shallowSetup() {
  // Sample props to pass to our shallow render
  const props = {
    book: { bookId: '', title: '', year: '', description: '' },
    match: {
      params: { id: '7ae5bfa3-f0d4-4fd3-8a9b-61676d67a3c8' },
    } as any,
    done: false,
    getBook: getBook,
    loading: false,
    history: {} as any,
    location: {} as any,
  };
  return props;
}
