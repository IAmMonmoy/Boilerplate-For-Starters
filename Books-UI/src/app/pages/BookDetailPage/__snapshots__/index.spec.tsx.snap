// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BookDeatilPage/> should match the snapshot 1`] = `
<Context.Provider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <Memo(Connect(BookDetailPage))
    book={
      Object {
        "bookId": "",
        "description": "",
        "title": "",
        "year": "",
      }
    }
    done={false}
    getBook={[MockFunction]}
    history={Object {}}
    loading={false}
    location={Object {}}
    match={
      Object {
        "params": Object {
          "id": "7ae5bfa3-f0d4-4fd3-8a9b-61676d67a3c8",
        },
      }
    }
  />
</Context.Provider>
`;
